from Bio.Application import _Option, _Switch, AbstractCommandline
class Primer3Commandline(_EmbossCommandLine):
    """Commandline object for the Primer3 interface from EMBOSS.

    The precise set of supported arguments depends on your version of EMBOSS.
    This version accepts arguments current at EMBOSS 6.1.0:

    >>> cline = Primer3Commandline(sequence="mysequence.fas", auto=True, hybridprobe=True)
    >>> cline.explainflag = True
    >>> cline.osizeopt=20
    >>> cline.psizeopt=200
    >>> cline.outfile = "myresults.out"
    >>> cline.bogusparameter = 1967  # Invalid parameter
    Traceback (most recent call last):
        ...
    ValueError: Option name bogusparameter was not found.
    >>> print(cline)
    eprimer3 -auto -outfile=myresults.out -sequence=mysequence.fas -hybridprobe=True -psizeopt=200 -osizeopt=20 -explainflag=True

    """

    def __init__(self, cmd='eprimer3', **kwargs):
        """Initialize the class."""
        self.parameters = [_Option(['-sequence', 'sequence'], 'Sequence to choose primers from.', is_required=True), _Option(['-task', 'task'], 'Tell eprimer3 what task to perform.'), _Option(['-hybridprobe', 'hybridprobe'], 'Find an internal oligo to use as a hyb probe.'), _Option(['-numreturn', 'numreturn'], 'Maximum number of primer pairs to return.'), _Option(['-includedregion', 'includedregion'], 'Subregion of the sequence in which to pick primers.'), _Option(['-target', 'target'], 'Sequence to target for flanking primers.'), _Option(['-excludedregion', 'excludedregion'], 'Regions to exclude from primer picking.'), _Option(['-forwardinput', 'forwardinput'], 'Sequence of a forward primer to check.'), _Option(['-reverseinput', 'reverseinput'], 'Sequence of a reverse primer to check.'), _Option(['-gcclamp', 'gcclamp'], "The required number of Gs and Cs at the 3' of each primer."), _Option(['-osize', 'osize'], 'Optimum length of a primer oligo.'), _Option(['-minsize', 'minsize'], 'Minimum length of a primer oligo.'), _Option(['-maxsize', 'maxsize'], 'Maximum length of a primer oligo.'), _Option(['-otm', 'otm'], 'Melting temperature for primer oligo (OBSOLETE).\n\nOption replaced in EMBOSS 6.6.0 by -opttm'), _Option(['-opttm', 'opttm'], 'Optimum melting temperature for a primer oligo.\n\nOption added in EMBOSS 6.6.0, replacing -otm'), _Option(['-mintm', 'mintm'], 'Minimum melting temperature for a primer oligo.'), _Option(['-maxtm', 'maxtm'], 'Maximum melting temperature for a primer oligo.'), _Option(['-maxdifftm', 'maxdifftm'], 'Maximum difference in melting temperatures between forward and reverse primers.'), _Option(['-ogcpercent', 'ogcpercent'], 'Optimum GC% for a primer.'), _Option(['-mingc', 'mingc'], 'Minimum GC% for a primer.'), _Option(['-maxgc', 'maxgc'], 'Maximum GC% for a primer.'), _Option(['-saltconc', 'saltconc'], 'Millimolar salt concentration in the PCR.'), _Option(['-dnaconc', 'dnaconc'], 'Nanomolar concentration of annealing oligos in the PCR.'), _Option(['-maxpolyx', 'maxpolyx'], 'Maximum allowable mononucleotide repeat length in a primer.'), _Option(['-psizeopt', 'psizeopt'], 'Optimum size for the PCR product.'), _Option(['-prange', 'prange'], 'Acceptable range of length for the PCR product.'), _Option(['-ptmopt', 'ptmopt'], 'Optimum melting temperature for the PCR product.'), _Option(['-ptmmin', 'ptmmin'], 'Minimum allowed melting temperature for the amplicon.'), _Option(['-ptmmax', 'ptmmax'], 'Maximum allowed melting temperature for the amplicon.'), _Option(['-oexcludedregion', 'oexcludedregion'], 'Do not pick internal oligos in this region.'), _Option(['-oligoinput', 'oligoinput'], 'Sequence of the internal oligo.'), _Option(['-osizeopt', 'osizeopt'], 'Optimum length of internal oligo.'), _Option(['-ominsize', 'ominsize'], 'Minimum length of internal oligo.'), _Option(['-omaxsize', 'omaxsize'], 'Maximum length of internal oligo.'), _Option(['-otmopt', 'otmopt'], 'Optimum melting temperature of internal oligo.'), _Option(['-otmmin', 'otmmin'], 'Minimum melting temperature of internal oligo.'), _Option(['-otmmax', 'otmmax'], 'Maximum melting temperature of internal oligo.'), _Option(['-ogcopt', 'ogcopt'], 'Optimum GC% for internal oligo.'), _Option(['-ogcmin', 'ogcmin'], 'Minimum GC% for internal oligo.'), _Option(['-ogcmax', 'ogcmax'], 'Maximum GC% for internal oligo.'), _Option(['-osaltconc', 'osaltconc'], 'Millimolar concentration of salt in the hybridisation.'), _Option(['-odnaconc', 'odnaconc'], 'Nanomolar concentration of internal oligo in the hybridisation.'), _Option(['-oanyself', 'oanyself'], 'Maximum allowable alignment score for self-complementarity.'), _Option(['-oendself', 'oendself'], "Max 3'-anchored self-complementarity global alignment score."), _Option(['-opolyxmax', 'opolyxmax'], 'Maximum length of mononucleotide repeat in internal oligo.'), _Option(['-mispriminglibraryfile', 'mispriminglibraryfile'], 'File containing library of sequences to avoid amplifying'), _Option(['-maxmispriming', 'maxmispriming'], 'Maximum allowed similarity of primers to sequences in library specified by -mispriminglibrary'), _Option(['-omishybmax', 'omishybmax'], 'Maximum alignment score for hybridisation of internal oligo to library specified by -mishyblibraryfile.'), _Option(['-mishyblibraryfile', 'mishyblibraryfile'], 'Library file of seqs to avoid internal oligo hybridisation.'), _Option(['-explainflag', 'explainflag'], 'Produce output tags with eprimer3 statistics')]
        _EmbossCommandLine.__init__(self, cmd, **kwargs)